File: AtCoder/KyoproTenkei/python/028-Cluttered_Paper-ans.py
class Rect:
    def __init__(self, lx, ly, rx, ry):
        self.lx = lx
        self.ly = ly
        self.rx = rx
        self.ry = ry


MAX_N = 100_000
MAX_X = MAX_Y = 1000
N = int(input())
rects = [Rect(*map(int, input().split())) for _ in range(N)]
cnt = [[0] * (MAX_X + 1) for _ in range(MAX_Y + 1)]
for r in rects:
    cnt[r.lx][r.ly] += 1
    cnt[r.lx][r.ry] -= 1
    cnt[r.rx][r.ly] -= 1
    cnt[r.rx][r.ry] += 1
for i in range(MAX_X + 1):
    for j in range(1, MAX_Y + 1):
        cnt[i][j] += cnt[i][j - 1]
for j in range(MAX_Y + 1):
    for i in range(1, MAX_X + 1):
        cnt[i][j] += cnt[i - 1][j]
A = [0] * (N + 1)
for i in range(MAX_X + 1):
    for j in range(MAX_Y + 1):
        if cnt[i][j] >= 1:
            A[cnt[i][j]] += 1
print('\n'.join(map(str, A[1:])))

File: AtCoder/KyoproTenkei/cpp/028-Cluttered_Paper-ans.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

const ll MAX_N = 100'000;
const ll MAX_X = 1000;
const ll MAX_Y = 1000;
ll N;
ll lxs[MAX_N], lys[MAX_N], rxs[MAX_N], rys[MAX_N];
ll cnt[MAX_X + 1][MAX_Y + 1];
ll A[MAX_N + 1];

int main() {
  cin >> N;
  for (ll i = 0; i < N; ++i) {
    cin >> lxs[i] >> lys[i] >> rxs[i] >> rys[i];
  }

  for (ll i = 0; i < N; ++i) {
    ++cnt[lxs[i]][lys[i]];
    --cnt[lxs[i]][rys[i]];
    --cnt[rxs[i]][lys[i]];
    ++cnt[rxs[i]][rys[i]];
  }
  for (ll i = 0; i <= MAX_X; ++i) {
    for (ll j = 1; j <= MAX_Y; ++j) {
      cnt[i][j] += cnt[i][j - 1];
    }
  }
  for (ll j = 0; j <= MAX_Y; ++j) {
    for (ll i = 1; i <= MAX_X; ++i) {
      cnt[i][j] += cnt[i - 1][j];
    }
  }
  for (ll i = 0; i <= MAX_X; ++i) {
    for (ll j = 0; j <= MAX_Y; ++j) {
      if (cnt[i][j] >= 1) ++A[cnt[i][j]];
    }
  }
  for (ll i = 1; i <= N; ++i) {
    printf("%lld\n", A[i]);
  }
}